### PR TITLE
[TCP] Inference mode batch_norm lowerings to TCP + tcp.const, tcp.sqrt ops

### PR DESCRIPTION
**Pull request by sjain-stanford**
Includes inference mode `torch.aten.batch_norm` decomposition to elementwise primitive ops (e.g. `tcp.mul`, `tcp.divf`, `tcp.sqrt`, `tcp.add`, `tcp.const`). The BN parameters are 1D tensors of size `[C]`, and are explicitly rank+shape broadcasted to ND tensors to match the input shape. 

Also includes ops and lowerings for `tcp.const` (contributed by @navahgar ) and `tcp.sqrt`.

### CODE REVIEW

File: externals/llvm-external-projects/torch-mlir-dialects/include/torch-mlir-dialects/Conversion/TcpToArith/TcpToArith.h
Comment (sanjoy) [inline externals/llvm-external-projects/torch-mlir-dialects/include/torch-mlir-dialects/Conversion/TcpToArith/TcpToArith.h:22]: I thought this decl was auto-generated in `torch-mlir-dialects/Conversion/Passes.h.inc`, can you please check?

File: externals/llvm-external-projects/torch-mlir-dialects/include/torch-mlir-dialects/Dialect/Tcp/IR/TcpOps.td
Comment (sanjoy) [inline externals/llvm-external-projects/torch-mlir-dialects/include/torch-mlir-dialects/Dialect/Tcp/IR/TcpOps.td:165]: Any constraints on the data types?

File: lib/Conversion/TorchToTcp/Elementwise.cpp
Comment (sanjoy) [inline lib/Conversion/TorchToTcp/Elementwise.cpp:125]: I believe MLIR coding convention is `runningMean`.  Same elsewhere.

File: lib/Conversion/TorchToTcp/Elementwise.cpp
Comment (sanjoy) [inline lib/Conversion/TorchToTcp/Elementwise.cpp:141]: Is this check in the verifier for `AtenBatchNormOp`?

File: lib/Conversion/TorchToTcp/Utils.cpp
Comment (sanjoy) [inline lib/Conversion/TorchToTcp/Utils.cpp:16]: Should this be `"mlir/IR/Value.h"`?

File: lib/Conversion/TorchToTcp/Utils.cpp
Comment (sanjoy) [inline lib/Conversion/TorchToTcp/Utils.cpp:166]: Again, MLIR convention is `numTotalElements` I believe.

File: lib/Conversion/TorchToTcp/Utils.cpp
Comment (sanjoy) [inline lib/Conversion/TorchToTcp/Utils.cpp:263]: I suggest extracting out a common helper and calling that with `DenseElementsAttr::get(const_type, vec)` from the three float/double/int variants.

**Review by sanjoy — CHANGES_REQUESTED** [review]

File: externals/llvm-external-projects/torch-mlir-dialects/include/torch-mlir-dialects/Conversion/TcpToArith/TcpToArith.h
Comment (navahgar) [reply externals/llvm-external-projects/torch-mlir-dialects/include/torch-mlir-dialects/Conversion/TcpToArith/TcpToArith.h:22]: Yes, I think that is possible now. But IIRC we decided to stick to the old style for now, to keep it uniform across all conversions in TorchMLIR.

File: lib/Conversion/TorchToTcp/Elementwise.cpp
Comment (navahgar) [inline lib/Conversion/TorchToTcp/Elementwise.cpp:264]: We are repeating this in every op now. We should probably extract this out into a helper function.

Not for this PR, probably do it as a separate cleanup PR.

File: test/Conversion/TorchToTcp/elementwise.mlir
Comment (navahgar) [inline test/Conversion/TorchToTcp/elementwise.mlir:160]: Can you add a test for lowering `ValueTensorLiteralOp`?

**Review by navahgar — COMMENTED** [review]
Can you run clang-format on the entire PR?

File: lib/Conversion/TorchToTcp/Elementwise.cpp
Comment (sjain-stanford) [reply lib/Conversion/TorchToTcp/Elementwise.cpp]: Fixed.

File: lib/Conversion/TorchToTcp/Elementwise.cpp
Comment (sjain-stanford) [reply lib/Conversion/TorchToTcp/Elementwise.cpp:141]: No AFAICT. `AtenBatchNormOp` [uses](https://sourcegraph.com/github.com/llvm/torch-mlir/-/blob/include/torch-mlir/Dialect/Torch/IR/GeneratedTorchOps.td?L4556-L4557) `AnyTorchOptionalTensorType` for running mean/variance. Depending on the mode (training/inference) the op is used in and whether it has been exposed to any training cycles at all, it is possible these are never initialized. At the moment we only support inference mode batch_norm lowering and when running stats are available, hence this check.

From PyTorch [docs](https://pytorch.org/docs/stable/generated/torch.nn.BatchNorm2d.html) for BatchNorm2d:
```
track_running_stats ([bool](https://docs.python.org/3/library/functions.html#bool)) – a boolean value that when set to True, this module tracks the running mean and variance, and when set to False, this module does not track such statistics, and initializes statistics buffers running_mean and running_var as None. When these buffers are None, this module always uses batch statistics, in both training and eval modes. Default: True
```

File: lib/Conversion/TorchToTcp/Utils.cpp
Comment (sjain-stanford) [reply lib/Conversion/TorchToTcp/Utils.cpp]: Good catch. This is not needed, and most likely got accidentally added by one of VSCode's plugins :)

Comment (sjain-stanford) [reply lib/Conversion/TorchToTcp/Utils.cpp]: Removed.

File: externals/llvm-external-projects/torch-mlir-dialects/include/torch-mlir-dialects/Dialect/Tcp/IR/TcpOps.td
Comment (sjain-stanford) [reply externals/llvm-external-projects/torch-mlir-dialects/include/torch-mlir-dialects/Dialect/Tcp/IR/TcpOps.td:165]: `tcp.const` op supports `Tcp_Tensor` type (floating point, signless integers, complex) for the result, however when lowering from TorchToTcp, for now we only support floating point cases (captured a TODO for future support of integer use-cases). Will clarify in the op's description.

File: test/Conversion/TorchToTcp/elementwise.mlir
Comment (sjain-stanford) [reply test/Conversion/TorchToTcp/elementwise.mlir:160]: This is unit-tested in the `torch.aten.batch_norm` testcase in `TorchToTcp/elementwise.mlir`. Did you want me to add a separate one anyway?

Comment (sjain-stanford) [reply test/Conversion/TorchToTcp/elementwise.mlir:160]: Added it to `misc.mlir`.

File: lib/Conversion/TorchToTcp/Utils.cpp
Comment (sjain-stanford) [reply lib/Conversion/TorchToTcp/Utils.cpp]: Good call. Done.

File: lib/Conversion/TorchToTcp/Elementwise.cpp
Comment (sjain-stanford) [reply lib/Conversion/TorchToTcp/Elementwise.cpp:264]: Sounds good. Let's do that in a separate PR.

File: lib/Conversion/TorchToTcp/Elementwise.cpp
Comment (navahgar) [reply lib/Conversion/TorchToTcp/Elementwise.cpp]: In that case, we should probably error out gracefully saying that is unsupported, instead of asserting here.

I think that is what @sanjoy was referring to as well - we can assert here if this was a check in the verifier of `AtenBatchNormOp`. If not, we should return an error instead.

File: lib/Conversion/TorchToTcp/Utils.h
Comment (sanjoy) [inline lib/Conversion/TorchToTcp/Utils.h:34]: Is this just called from `getConstTensor`?  If so, it should be in an `impl` namespace to make it clear that it is a helper.

File: lib/Conversion/TorchToTcp/Utils.h
Comment (sanjoy) [inline lib/Conversion/TorchToTcp/Utils.h:38]: Add docs.

File: lib/Conversion/TorchToTcp/Utils.h
Comment (sanjoy) [inline lib/Conversion/TorchToTcp/Utils.h:30]: These and earlier functions should have comments (maybe just one combined comment on the top).

File: lib/Conversion/TorchToTcp/Utils.cpp
Comment (sanjoy) [inline lib/Conversion/TorchToTcp/Utils.cpp:88]: LLVM convention is to add a short message to the assertion like:

```
assert(inputType.getRank() == 0 && "Only scalars are supported!");
```

File: lib/Conversion/TorchToTcp/Elementwise.cpp
Comment (sjain-stanford) [reply lib/Conversion/TorchToTcp/Elementwise.cpp]: Ah makes sense - will switch it to a `rewriter.notifyMatchFailure` here.

File: lib/Conversion/TorchToTcp/Utils.h
Comment (sanjoy) [inline lib/Conversion/TorchToTcp/Utils.h:37]: I prefer these have unit tests too, but OK to do in a later PR.

**Review by sanjoy — APPROVED** [review]
LGTM modulo minor comments

File: lib/Conversion/TorchToTcp/Elementwise.cpp
Comment (ramiro050) [inline lib/Conversion/TorchToTcp/Elementwise.cpp:111]: nit: I don't think this line is needed

File: lib/Conversion/TorchToTcp/Elementwise.cpp
Comment (ramiro050) [inline lib/Conversion/TorchToTcp/Elementwise.cpp:136]: nit: I don't think the `.template` is needed

File: lib/Conversion/TorchToTcp/Elementwise.cpp
Comment (ramiro050) [inline lib/Conversion/TorchToTcp/Elementwise.cpp]: This *can* happen in valid torch dialect MLIR, right? If so, an assert should not be used. I would use a `notifyMatchFailure`

File: lib/Conversion/TorchToTcp/Elementwise.cpp
Comment (ramiro050) [inline lib/Conversion/TorchToTcp/Elementwise.cpp:163]: Should there be a period after `vectors`?

File: lib/Conversion/TorchToTcp/Elementwise.cpp
Comment (ramiro050) [inline lib/Conversion/TorchToTcp/Elementwise.cpp:114]: This ops takes two extra arguments, `momentum` and `training` that are currently not being handled here.

File: lib/Conversion/TorchToTcp/Elementwise.cpp
Comment (sjain-stanford) [reply lib/Conversion/TorchToTcp/Elementwise.cpp]: Done

File: lib/Conversion/TorchToTcp/Utils.cpp
Comment (ramiro050) [inline lib/Conversion/TorchToTcp/Utils.cpp:119]: The implementation of the 1D and 0D case are almost identical. All that is needed is two `if` statements conditioned on the input rank to create the reassociations and the `dimSizes`. I would refactor all the logic into a helper function.

File: lib/Conversion/TorchToTcp/Utils.cpp
Comment (ramiro050) [inline lib/Conversion/TorchToTcp/Utils.cpp:169]: Should there be a check here that the `shape` has only static sizes?

File: externals/llvm-external-projects/torch-mlir-dialects/lib/Dialect/Tcp/IR/TcpOps.cpp
Comment (ramiro050) [inline externals/llvm-external-projects/torch-mlir-dialects/lib/Dialect/Tcp/IR/TcpOps.cpp:69]: Is there a test for this fold?

**Review by ramiro050 — CHANGES_REQUESTED** [review]

File: externals/llvm-external-projects/torch-mlir-dialects/include/torch-mlir-dialects/Conversion/TcpToArith/TcpToArith.h
Comment (navahgar) [reply externals/llvm-external-projects/torch-mlir-dialects/include/torch-mlir-dialects/Conversion/TcpToArith/TcpToArith.h:22]: @ramiro050 I think this is the one: https://discourse.llvm.org/t/psa-deprecation-warning-per-pass-autogenerated-macros-and-pass-options/64998

File: lib/Conversion/TorchToTcp/Elementwise.cpp
Comment (sjain-stanford) [reply lib/Conversion/TorchToTcp/Elementwise.cpp]: Thanks, yes I fixed this from an earlier comment but it probably didn't update while you were reviewing. This is fixed now.

File: lib/Conversion/TorchToTcp/Utils.h
Comment (sjain-stanford) [reply lib/Conversion/TorchToTcp/Utils.h]: Took a TODO to add this (cc: @navahgar).

**Comment by ramiro050** [comment]
Not sure if this was discussed in a previous PR, but should there be a CI for merges into the mlir-tcp branch?

File: lib/Conversion/TorchToTcp/Misc.cpp
Comment (sjain-stanford) [reply lib/Conversion/TorchToTcp/Misc.cpp]: Good point. Dropped the inner `if`: 
```
    if (auto elements = op.getValueAttr().dyn_cast<DenseIntElementsAttr>())
      return rewriter.notifyMatchFailure(
          op, "Integer constants lowering to TCP is unimplemented");
```

**Comment by sjain-stanford** [comment]
> Not sure if this was discussed in a previous PR, but should there be a CI for merges into the mlir-tcp branch?

This would be great actually. We didn't do it yet to avoid burdening the GHA workers with TCP flows, but if this is not a concern, I will (in a follow-on PR) update the GHA workflow to also enable TCP for all PRs to `mlir-tcp` branch.

**Comment by ramiro050** [comment]
> This would be great actually. We didn't do it yet to avoid burdening the GHA workers with TCP flows, but if this is not a concern, I will (in a follow-on PR) update the GHA workflow to also enable TCP for all PRs to `mlir-tcp` branch.

@silvasean @powderluv, are there any GHA concerns?

File: lib/Conversion/TorchToTcp/Elementwise.cpp
Comment (sjain-stanford) [reply lib/Conversion/TorchToTcp/Elementwise.cpp]: Thanks, fixed!

File: lib/Conversion/TorchToTcp/Utils.cpp
Comment (sjain-stanford) [reply lib/Conversion/TorchToTcp/Utils.cpp:175]: Good point. Done, PTAL.
```
  for (int64_t a : shape) {
    if (a == mlir::torch::Torch::kUnknownSize) {
        op->emitOpError("getConstTensor(): only static sizes supported");
        return std::nullopt;
    }
    numTotalElements *= a;
  }
```

**Comment by silvasean** [comment]
As long as it doesn't significantly impact the rest of the project I'm fine with it -- but how do we measure if it impacts the rest of the project? Also it would be nice if we can avoid getting emails from the tcp branch.

File: lib/Conversion/TorchToTcp/Elementwise.cpp
Comment (sjain-stanford) [reply lib/Conversion/TorchToTcp/Elementwise.cpp:111]: Done, PTAL. We only support `training=false`, and ignore `momentum` and `cudnnEnabled`.
```c++
    // momentum is ignored
    Value momentum = adaptor.getMomentum();
    (void)momentum;

    // cudnnEnabled is ignored
    Value cudnnEnabled = adaptor.getCudnnEnabled();
    (void)cudnnEnabled;

    bool training = false;
    if (!matchPattern(op.getTraining(), m_TorchConstantBool(&training)))
      return rewriter.notifyMatchFailure(
        op, "non-bool training unsupported");
    if (training)
      return rewriter.notifyMatchFailure(
        op, "only inference mode batch_norm lowering supported");
```

File: lib/Conversion/TorchToTcp/Utils.cpp
Comment (ramiro050) [reply lib/Conversion/TorchToTcp/Utils.cpp:175]: I think assuming that `shape` was created from `Torch` shapes is fragile. At the level of the `TorchToTcp` pass, there will be torch types and builtin tensor types in use. I would instead simply check that `a` is positive.

File: externals/llvm-external-projects/torch-mlir-dialects/include/torch-mlir-dialects/Dialect/Tcp/IR/TcpOps.td
Comment (sanjoy) [inline externals/llvm-external-projects/torch-mlir-dialects/include/torch-mlir-dialects/Dialect/Tcp/IR/TcpOps.td:166]: Nit: I don't think the `that are used in other ops` adds anything -- I'd just drop it.

File: lib/Conversion/TorchToTcp/Utils.cpp
Comment (sjain-stanford) [reply lib/Conversion/TorchToTcp/Utils.cpp:171]: Replaced with an assert, since this utility creates `tcp.const` and should never be called with dynamic shapes. It is the caller's responsibility to ensure this.
```
assert(a >= 0 && "getConstTensor(): Only static shapes supported");
```

File: lib/Conversion/TorchToTcp/Utils.cpp
Comment (sjain-stanford) [reply lib/Conversion/TorchToTcp/Utils.cpp]: Done, PTAL.

File: externals/llvm-external-projects/torch-mlir-dialects/lib/Dialect/Tcp/IR/TcpOps.cpp
Comment (sjain-stanford) [reply externals/llvm-external-projects/torch-mlir-dialects/lib/Dialect/Tcp/IR/TcpOps.cpp:69]: I encountered some issues when testing the constant folding. @navahgar (who authored this op) said he will be able to look into this sometime early next week. I believe this is the last open comment before landing the PR. In the meantime, please feel free to re-review the other changes.

File: lib/Conversion/TorchToTcp/Misc.cpp
Comment (ramiro050) [inline lib/Conversion/TorchToTcp/Misc.cpp:84]: nit: not needed

**Review by ramiro050 — APPROVED** [review]
LGTM!

File: externals/llvm-external-projects/torch-mlir-dialects/lib/Dialect/Tcp/IR/TcpOps.cpp
Comment (sjain-stanford) [reply externals/llvm-external-projects/torch-mlir-dialects/lib/Dialect/Tcp/IR/TcpOps.cpp]: Per discussion with @navahgar , removed `hasFolder` from `Tcp_ConstOp` tablegen description, and `ConstOp::fold` for now. Noted a TODO to re-enable folding along with unit-tests.

File: lib/Conversion/TorchToTcp/Misc.cpp
Comment (sjain-stanford) [reply lib/Conversion/TorchToTcp/Misc.cpp]: Last one remaining. Thanks!

File: externals/llvm-external-projects/torch-mlir-dialects/lib/Dialect/Tcp/IR/TcpOps.cpp
Comment (ramiro050) [reply externals/llvm-external-projects/torch-mlir-dialects/lib/Dialect/Tcp/IR/TcpOps.cpp]: Great!

**Pull request merged**
